pyramid/session.py

```diff
@@ -101,6 +108,77 @@ def signed_deserialize(serialized, secret, hmac=hmac):
 
     return pickle.loads(pickled)
 
+
+def check_csrf_origin(request):
+    """
+    Check the Origin of the request to see if it is a cross site request or
+    not.
+
+    Note that this function will do nothing if request.scheme is not https.
+
+    .. versionadded:: 1.7
+    """
+    if request.scheme == "https":
+        # Suppose user visits http://example.com/
+        # An active network attacker (man-in-the-middle, MITM) sends a
+        # POST form that targets https://example.com/detonate-bomb/ and
+        # submits it via JavaScript.
+        #
+        # The attacker will need to provide a CSRF cookie and token, but
+        # that's no problem for a MITM when we cannot make any assumptions
+        # about what kind of session storage is being used. So the MITM can
+        # circumvent the CSRF protection. This is true for any HTTP connection,
+        # but anyone using HTTPS expects better! For this reason, for
+        # https://example.com/ we need additional protection that treats
+        # http://example.com/ as completely untrusted. Under HTTPS,
+        # Barth et al. found that the Referer header is missing for
+        # same-domain requests in only about 0.2% of cases or less, so
+        # we can use strict Referer checking.
+
+        # Determine the origin of this request
+        origin = request.headers.get("Origin")
+        if origin is None:
+            origin = request.referrer
+
+        # Fail if we were not able to locate an origin at all
+        if not origin:
+            raise BadCSRFOrigin(
+                "Origin checking failed - no Origin or Referer."
+            )
+
+        # Parse our origin so we we can extract the required information from
+        # it.
+        originp = urlparse.urlparse(origin)
+
+        # Ensure that our Referer is also secure.
+        if originp.scheme != "https":
+            raise BadCSRFOrigin(
+                "Referer checking failed - Referer is insecure while host is "
+                "secure."
+            )
+
+        # Determine which origins we trust, which by default will include the
+        # current origin.
+        trusted_origins = request.registry.settings.get(
```

Comment:
Where is this being cast from an ini string syntax to a list? I can't find it if it's being done right now. This probably needs to be wrapped in `pyramid.settings.aslist`.